src/microbe_stage/MicrobeAI.cs

```diff
@@ -369,44 +396,94 @@ private void RunAndTumble(Random random)
-        {
-            compoundDifference += compound.Value;
+            float quantityAbsorbedThisStep, quantityAbsorbedPreviousStep;
+            microbe.TotalAbsorbedCompounds.TryGetValue(compoundWeight.Key, out quantityAbsorbedThisStep);
+            previouslyAbsorbedCompounds.TryGetValue(compoundWeight.Key, out quantityAbsorbedPreviousStep);
+
+            compoundDifference += quantityAbsorbedThisStep - quantityAbsorbedPreviousStep;
+
+            compoundDifference *= compoundWeight.Value;
+            gradientValue += compoundDifference;
         }
 
+        // Implement a detection threshold to possibly rule out too tiny variations
+        // TODO: possibly include cell capacity correction
+        float differenceDetectionThreshold = Constants.AI_GRADIENT_DETECTION_THRESHOLD;
+
         // If food density is going down, back up and see if there's some more
-        if (compoundDifference < 0 && random.Next(0, 10) < 9)
+        if (gradientValue < -differenceDetectionThreshold && random.Next(0, 10) < 9)
         {
             MoveWithRandomTurn(2.5f, 3.0f, random);
         }
 
         // If there isn't any food here, it's a good idea to keep moving
-        if (compoundDifference == 0 && random.Next(0, 10) < 5)
+        if (Math.Abs(gradientValue) <= differenceDetectionThreshold && random.Next(0, 10) < 5)
         {
             MoveWithRandomTurn(0.0f, 0.4f, random);
         }
 
-        // If positive last step you gained compounds, so let's stick around
-        if (compoundDifference > 0)
+        // If positive last step you gained compounds, so let's move toward the source
+        if (gradientValue > differenceDetectionThreshold)
         {
-            // There's a decent chance to turn most of the way around
+            // There's a decent chance to turn by 90° to explore gradient
+            // 180° is useless since previous position let you absorb less compounds already
             if (random.Next(0, 10) < 4)
             {
-                MoveWithRandomTurn(0.0f, 3.0f, random);
+                MoveWithRandomTurn(0.0f, 1.5f, random);
             }
         }
     }
 
+    /// <summary>
+    ///   Prioritizing compounds that are stored in lesser quantities.
+    ///   If ATP-producing compounds are low (less than half storage capacities),
+    ///   non ATP-related compounds are discarded.
+    ///   Updates compoundsSearchWeights class dictionary.
+    /// </summary>
+    private void ComputeCompoundsSearchWeights()
+    {
+        IEnumerable<Compound> usefulCompounds = microbe.TotalAbsorbedCompounds.Keys;
+
+        // If this microbe lacks vital compounds don't bother with ammonia and phosphate
+        if (microbe.TotalAbsorbedCompounds.Keys.Any(
+            compound => IsVitalCompound(compound) &&
+                microbe.Compounds.GetCompoundAmount(compound) < 0.5f * microbe.Compounds.Capacity))
+        {
+            usefulCompounds = usefulCompounds.Where(x => x != ammonia && x != phosphates);
```

Comment:
I'd use `usefulCompounds = usefulCompounds.Except(new[] { ammonia, phosphates });`

Reply:
just noticed that that isn't really your code but adopted :D
But still, I'd prefer an `Except`-usage here

Reply:
Yes it was adopted, but I'm definitely in favor of using `Except` as well, now that I know of it. :)

Reply:
Well the Except syntax shown above, probably allocates a new two element array. Just to remind everyone that we are working on a realtime game and not https://github.com/Revolutionary-Games/ThriveDevCenter where I would allow this kind of code in 0 milliseconds.